### PR TITLE
A test for measure the performance effect of vectorized intrinsic on fp16 to elementwise op

### PR DESCRIPTION
**Pull request by JamesLim-sy**
### PR types

Performance optimization
### PR changes

OPs
### Describe

- A try for `__hadd2 intrinsic` , but kernel performance optimization seems little (no more than 0.4% increase for typical ternary add cases in AlphaFold model.)

~~- Once finishing the performance optimization for broadcast calculation last summer in [PR32512](https://github.com/PaddlePaddle/Paddle/pull/32512), a bunch of code porting has been estabulished here, and the performance fluctuats consequently. This pr is to improve the performance of special broadcast binary or tenery case below :~~
~~- x_shape: [16, 1, 513, 513], fp16~~
~~- y_shape: [1], fp16~~

~~- Performance:~~

~~-  More work for performance fix will be implemented later.~~

~~- **Make `broadcast kernel` as the common implement of `elementwise&broadcast calculation` leaves little space for performance optimization, duing to that time cost of `elementwise calculation` is much smaller than `broadcast calculation`, so any change in `broadcast kernel` may slightly increase the time cost of `elementwise` .  Before this  change fixed, this pr shall shut down.**~~

### CODE REVIEW

File: paddle/phi/kernels/funcs/broadcast_function.h [hunk 1]
```diff
@@ -27,6 +27,17 @@ namespace funcs {
 
 #if defined(__NVCC__) || defined(__HIPCC__) || defined(__xpu__)
 
+enum BroadcastType {
+  kNoUse = 0,   // no broadcast
```
Comment (Xreki) [inline paddle/phi/kernels/funcs/broadcast_function.h]: 这三个命名可以再考虑下，kNoBroadcast、kCommon、kOneToMany？可以下个PR再改。

File: paddle/phi/kernels/primitive/datamover_primitives.h [hunk 2]
```diff
@@ -793,6 +793,27 @@ __device__ __forceinline__ void ReadDataBc(
   }
 }
 
+template <typename T, int NX, int NY = 1, bool IsBoundary = false>
+__device__ __forceinline__ void ReadOneData(T* dst,
```
Comment (Xreki) [inline paddle/phi/kernels/primitive/datamover_primitives.h]: 这个函数的命名可以再考虑下。

**Review by Xreki — DISMISSED** [review]
LGTM and Great work~ 可以请 @niuliling123 review下。

**Comment by AnnaTrainingG** [comment]
KP CI 错误，不能合入代码

File: paddle/phi/kernels/primitive/datamover_primitives.h [hunk 3]
```diff
@@ -793,6 +793,27 @@ __device__ __forceinline__ void ReadDataBc(
...
+  T src_data = src[0];
+
+#pragma unroll
+  for (uint32_t nx = 0; nx < NX; ++nx) {
+    uint32_t index_output = thread_offset + nx;
+    if (IsBoundary) {
+      if (index_output >= total_num_output) {
+        break;
+      }
+    }
```
Comment (AnnaTrainingG) [inline paddle/phi/kernels/primitive/datamover_primitives.h]: 这一大段都是在计算index_output 但是你并没有用到，这里需要删除

File: paddle/phi/kernels/primitive/datamover_primitives_xpu2.h [hunk 4]
```diff
@@ -1211,6 +1211,27 @@ __device__ __inline__ void ReadDataBc(T* dst,
...
+  T src_data = src[0];
+
+#pragma unroll
+  for (uint32_t nx = 0; nx < NX; ++nx) {
+    uint32_t index_output = thread_offset + nx;
+    if (IsBoundary) {
+      if (index_output >= total_num_output) {
+        break;
+      }
+    }
```
Comment (AnnaTrainingG) [inline paddle/phi/kernels/primitive/datamover_primitives_xpu2.h]: 同上

File: paddle/phi/kernels/primitive/datamover_primitives_xpu2.h [hunk 5]
```diff
@@ -1211,6 +1211,27 @@ __device__ __inline__ void ReadDataBc(T* dst,
...
 }
 
+template <typename T, int NX, int NY = 1, bool IsBoundary = false>
+__device__ __forceinline__ void ReadOneData(T* dst,
+                                            const T* __restrict__ src,
+                                            uint32_t block_offset,
+                                            int total_num_output,
+                                            int read_lens = NX) {
+  uint32_t thread_offset = block_offset + threadIdx.x * NX;
+  T src_data = src[0];
```
Comment (AnnaTrainingG) [inline paddle/phi/kernels/primitive/datamover_primitives_xpu2.h]: XPU不支持这样从GM读取数据，需要使用GM2LM

File: paddle/phi/kernels/primitive/datamover_primitives_xpu2.h [hunk 6]
```diff
@@ -1211,6 +1211,16 @@ __device__ __inline__ void ReadDataBc(T* dst,
   }
 }
 
+template <typename T, int NX>
+__device__ __inline__ void ReadOneToMany(T* dst, T* src) {
```
Comment (AnnaTrainingG) [inline paddle/phi/kernels/primitive/datamover_primitives_xpu2.h]: 这里的src 需要用global_ptr修饰，建议参考前面的

Comment (JamesLim-sy) [reply paddle/phi/kernels/primitive/datamover_primitives_xpu2.h]: 根据建议修改

File: paddle/phi/kernels/primitive/datamover_primitives.h (hunk 2 above)
Comment (JamesLim-sy) [reply paddle/phi/kernels/primitive/datamover_primitives.h]: 根据建议修改

File: paddle/phi/kernels/funcs/broadcast_function.h (hunk 1 above)
Comment (JamesLim-sy) [reply paddle/phi/kernels/funcs/broadcast_function.h]: 根据建议修改

File: paddle/phi/kernels/primitive/datamover_primitives.h (hunk 3 above)
Comment (JamesLim-sy) [reply paddle/phi/kernels/primitive/datamover_primitives.h]: 已根据建议删除

File: paddle/phi/kernels/primitive/datamover_primitives_xpu2.h (hunk 4 above)
Comment (JamesLim-sy) [reply paddle/phi/kernels/primitive/datamover_primitives_xpu2.h]: 已根据建议删除

**Review by bmb0537 — DISMISSED** [review]

**Pull request closed without merging**